File: rust/examples/basic_usage.rs
//! Basic usage example for links-mq.
//!
//! This example demonstrates the core Link and LinkStore interfaces.
//!
//! Run with: `cargo run --example basic_usage`

use links_mq::{Any, Link, LinkPattern, LinkRef};

fn main() {
    // =========================================================================
    // Creating Links
    // =========================================================================

    println!("=== Creating Links ===\n");

    // Simple link with u64 IDs
    let simple_link = Link::new(1u64, LinkRef::Id(2), LinkRef::Id(3));
    println!("Simple link: {:?}", simple_link);

    // Link with u32 IDs (smaller memory footprint)
    let u32_link = Link::new(1u32, LinkRef::Id(2u32), LinkRef::Id(3u32));
    println!("u32 link: {:?}", u32_link);

    // Link with usize IDs (platform-native size)
    let usize_link = Link::new(1usize, LinkRef::Id(2usize), LinkRef::Id(3usize));
    println!("usize link: {:?}", usize_link);

    // =========================================================================
    // Nested Links (Recursive Structures)
    // =========================================================================

    println!("\n=== Nested Links ===\n");

    // Create nested link structures for complex relationships
    let inner = Link::new(10u64, LinkRef::Id(20), LinkRef::Id(30));
    let outer = Link::new(1u64, LinkRef::link(inner.clone()), LinkRef::Id(5));
    println!("Outer link with nested source: {:?}", outer);
    println!("  Outer source ID: {}", outer.source_id()); // 10

    // Deep nesting: (1: (2: (3: 30 31)))
    let level3 = Link::new(3u64, LinkRef::Id(30), LinkRef::Id(31));
    let level2 = Link::new(2u64, LinkRef::link(level3), LinkRef::Id(21));
    let level1 = Link::new(1u64, LinkRef::link(level2), LinkRef::Id(11));
    println!("Deeply nested: {:?}", level1);
    println!("  Level 1 source ID: {}", level1.source_id()); // 2

    // =========================================================================
    // Universal Links (Variable Number of References)
    // =========================================================================

    println!("\n=== Universal Links ===\n");

    // Universal link with additional values beyond source/target
    let universal = Link::with_values(
        100u64,
        LinkRef::Id(1), // subject
        LinkRef::Id(2), // predicate
        vec![LinkRef::Id(3), LinkRef::Id(4), LinkRef::Id(5)], // objects
    );
    println!("Universal link: {:?}", universal);
    println!("  Has values: {}", universal.has_values());

    // =========================================================================
    // Special Link Types
    // =========================================================================

    println!("\n=== Special Link Types ===\n");

    // Point link (self-referential: id == source == target)
    let point = Link::<u64>::point(42);
    println!("Point link: {:?}", point);
    println!("  Is point: {}", point.is_point()); // true

    // Null/nothing link (all zeros)
    let null_link = Link::<u64>::nothing();
    println!("Null link: {:?}", null_link);
    println!("  Is null: {}", null_link.is_null()); // true

    // =========================================================================
    // LinkRef Operations
    // =========================================================================

    println!("\n=== LinkRef Operations ===\n");

    // Create refs from IDs
    let ref_id: LinkRef<u64> = LinkRef::Id(42);
    println!("ID ref: {:?}", ref_id);
    println!("  Is ID: {}", ref_id.is_id()); // true
    println!("  Get ID: {}", ref_id.get_id()); // 42

    // Create refs from links
    let ref_link: LinkRef<u64> = LinkRef::link(simple_link.clone());
    println!("Link ref: {:?}", ref_link);
    println!("  Is link: {}", ref_link.is_link()); // true
    println!("  Get ID: {}", ref_link.get_id()); // 1 (the link's ID)

    // Using From trait
    let from_id: LinkRef<u64> = 99u64.into();
    let from_link: LinkRef<u64> = simple_link.clone().into();
    println!("From ID: {:?}", from_id);
    println!("From link: {:?}", from_link);

    // =========================================================================
    // Pattern Matching
    // =========================================================================

    println!("\n=== Pattern Matching ===\n");

    let links = vec![
        Link::new(1u64, LinkRef::Id(10), LinkRef::Id(20)),
        Link::new(2u64, LinkRef::Id(10), LinkRef::Id(30)),
        Link::new(3u64, LinkRef::Id(20), LinkRef::Id(30)),
        Link::new(4u64, LinkRef::Id(30), LinkRef::Id(40)),
    ];

    // Match all links
    let all_pattern = LinkPattern::<u64>::all();
    let all_matched: Vec<_> = links.iter().filter(|l| all_pattern.matches(l)).collect();
    println!("All links count: {}", all_matched.len()); // 4

    // Match by source
    let source_pattern = LinkPattern::with_source(LinkRef::Id(10u64));
    let with_source_10: Vec<_> = links
        .iter()
        .filter(|l| source_pattern.matches(l))
        .map(|l| l.id)
        .collect();
    println!("Links with source=10: {:?}", with_source_10); // [1, 2]

    // Match by target
    let target_pattern = LinkPattern::with_target(LinkRef::Id(30u64));
    let with_target_30: Vec<_> = links
        .iter()
        .filter(|l| target_pattern.matches(l))
        .map(|l| l.id)
        .collect();
    println!("Links with target=30: {:?}", with_target_30); // [2, 3]

    // Match with Any wildcard (builder pattern)
    let any_source_pattern = LinkPattern::<u64>::new()
        .source(Any)
        .target(30u64);
    let any_source_target_30: Vec<_> = links
        .iter()
        .filter(|l| any_source_pattern.matches(l))
        .map(|l| l.id)
        .collect();
    println!("Links with Any source, target=30: {:?}", any_source_target_30); // [2, 3]

    // Match by source AND target
    let exact_pattern = LinkPattern::with_source_target(
        LinkRef::Id(10u64),
        LinkRef::Id(20u64),
    );
    let exact_match: Vec<_> = links
        .iter()
        .filter(|l| exact_pattern.matches(l))
        .map(|l| l.id)
        .collect();
    println!("Links with source=10, target=20: {:?}", exact_match); // [1]

    println!("\n=== Done ===\n");
}

File: js/examples/basic-usage.js
/**
 * Basic usage examples for links-queue
 *
 * This file demonstrates the core Link and LinkStore interfaces.
 *
 * Run with any runtime:
 * - Node.js: node examples/basic-usage.js
 * - Bun: bun examples/basic-usage.js
 * - Deno: deno run examples/basic-usage.js
 */

import {
  Any,
  isLink,
  isLinkId,
  isLinkRef,
  getLinkId,
  createLink,
  matchesPattern,
} from '../src/index.js';

// =============================================================================
// Creating Links
// =============================================================================

console.log('=== Creating Links ===\n');

// Simple link with numeric IDs
const simpleLink = createLink(1, 2, 3);
console.log('Simple link:', simpleLink);
// Output: { id: 1, source: 2, target: 3 }

// Link with bigint IDs (for large datasets)
const bigintLink = createLink(1n, 9007199254740993n, 9007199254740994n);
console.log('Bigint link:', bigintLink);

// Link with string IDs (for UUID-based systems)
const uuidLink = createLink('link-001', 'source-uuid-123', 'target-uuid-456');
console.log('UUID link:', uuidLink);

// =============================================================================
// Nested Links (Recursive Structures)
// =============================================================================

console.log('\n=== Nested Links ===\n');

// Create nested link structures for representing complex relationships
const personType = createLink(10, 0, 0); // A "type" link
const nameProperty = createLink(11, 0, 0); // A "name" property link
const person = createLink(1, personType, nameProperty);
console.log('Person link with type and name:', person);

// Deep nesting: (1: (2: (3: 30 31)))
const level3 = createLink(3, 30, 31);
const level2 = createLink(2, level3, 21);
const level1 = createLink(1, level2, 11);
console.log('Deeply nested:', level1);
console.log('  Level 1 source ID:', level1.source.id); // 2
console.log('  Level 2 source ID:', level1.source.source.id); // 3

// =============================================================================
// Universal Links (Variable Number of References)
// =============================================================================

console.log('\n=== Universal Links ===\n');

// Universal link with additional values beyond source/target
// Useful for n-ary relationships
const tripleRelation = createLink(
  100, // id
  1, // source (subject)
  2, // target (predicate)
  [3, 4, 5] // values (objects)
);
console.log('Triple+ relation:', tripleRelation);

// =============================================================================
// Type Checking
// =============================================================================

console.log('\n=== Type Checking ===\n');

console.log(
  'isLink({ id: 1, source: 2, target: 3 }):',
  isLink({ id: 1, source: 2, target: 3 })
); // true
console.log('isLink(42):', isLink(42)); // false
console.log('isLink(null):', isLink(null)); // false

console.log('isLinkId(42):', isLinkId(42)); // true
console.log('isLinkId(1n):', isLinkId(1n)); // true
console.log('isLinkId("uuid"):', isLinkId('uuid')); // true
console.log('isLinkId({}):', isLinkId({})); // false

console.log('isLinkRef(42):', isLinkRef(42)); // true
console.log('isLinkRef(simpleLink):', isLinkRef(simpleLink)); // true

// =============================================================================
// Extracting IDs
// =============================================================================

console.log('\n=== Extracting IDs ===\n');

console.log('getLinkId(42):', getLinkId(42)); // 42
console.log('getLinkId(simpleLink):', getLinkId(simpleLink)); // 1

const nestedSource = createLink(5, 50, 51);
const linkWithNestedSource = createLink(1, nestedSource, 10);
console.log(
  'getLinkId(linkWithNestedSource.source):',
  getLinkId(linkWithNestedSource.source)
); // 5

// =============================================================================
// Pattern Matching
// =============================================================================

console.log('\n=== Pattern Matching ===\n');

const links = [
  createLink(1, 10, 20),
  createLink(2, 10, 30),
  createLink(3, 20, 30),
  createLink(4, 30, 40),
];

// Find links with source = 10
const sourcePattern = { source: 10 };
const withSource10 = links.filter((link) =>
  matchesPattern(link, sourcePattern)
);
console.log(
  'Links with source=10:',
  withSource10.map((l) => l.id)
); // [1, 2]

// Find links with target = 30
const targetPattern = { target: 30 };
const withTarget30 = links.filter((link) =>
  matchesPattern(link, targetPattern)
);
console.log(
  'Links with target=30:',
  withTarget30.map((l) => l.id)
); // [2, 3]

// Find links with any source and target = 30
const anySourcePattern = { source: Any, target: 30 };
const anySourceTarget30 = links.filter((link) =>
  matchesPattern(link, anySourcePattern)
);
console.log(
  'Links with Any source, target=30:',
  anySourceTarget30.map((l) => l.id)
); // [2, 3]

// Match all links
const allPattern = { source: Any, target: Any };
const allMatched = links.filter((link) => matchesPattern(link, allPattern));
console.log('All links (Any, Any):', allMatched.length); // 4

console.log('\n=== Done ===\n');
